google/cloud/spanner/internal/connection_impl.h

```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_CONNECTION_IMPL_H_
+#define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_CONNECTION_IMPL_H_
+
+#include "google/cloud/spanner/connection.h"
+#include "google/cloud/spanner/internal/spanner_stub.h"
+#include "google/cloud/spanner/version.h"
+#include "google/cloud/status_or.h"
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+// A concrete `Connection` subclass that uses gRPC to actually talk to a real
+// Spanner instance. See `MakeConnection()` for a factory function that creates
+// and returns instances of this class.
+class ConnectionImpl : public Connection {
+ public:
+  // Creates a ConnectionImpl that will talk to the specified `database` using
+  // the given `stub`. We can test this class by injecting in a mock `stub`.
+  explicit ConnectionImpl(std::string database,
+                          std::shared_ptr<internal::SpannerStub> stub)
+      : database_(std::move(database)), stub_(std::move(stub)) {}
+
+  StatusOr<CommitResult> Commit(CommitParams) override;
```

Comment:
Should we use `final`?

Reply:
> Don't know. Maybe? I don't have a sense of when to apply `final`. Suggestion?

My only suggestions would be to forget about it, or apply it to every override (unless that contradicts some explicit need).  I'm happy with the former for now.